library/src/main/java/org/apache/apex/malhar/lib/window/impl/SpillableWindowedKeyedStorage.java

```diff
@@ -0,0 +1,198 @@
+  private long bucket;
+  @NotNull
+  private String identifier;
+  @NotNull
+  private Serde<Window, Slice> windowSerde;
+  @NotNull
+  private Serde<ImmutablePair<Window, K>, Slice> windowKeyPairSerde;
+  @NotNull
+  private Serde<K, Slice> keySerde;
+  @NotNull
+  private Serde<V, Slice> valueSerde;
+
+  protected Spillable.SpillableByteMap<ImmutablePair<Window, K>, V> internValues;
+  protected Spillable.SpillableByteArrayListMultimap<Window, K> internKeys;
+
+  public SpillableWindowedKeyedStorage()
+  {
+  }
+
+  public SpillableWindowedKeyedStorage(long bucket, String identifier,
+      Serde<Window, Slice> windowSerde, Serde<ImmutablePair<Window, K>, Slice> windowKeyPairSerde, Serde<K, Slice> keySerde, Serde<V, Slice> valueSerde)
+  {
+    this.bucket = bucket;
+    this.identifier = identifier;
+    this.windowSerde = windowSerde;
+    this.windowKeyPairSerde = windowKeyPairSerde;
+    this.keySerde = keySerde;
+    this.valueSerde = valueSerde;
+  }
+
+  public void setStore(SpillableStateStore store)
+  {
+    this.store = store;
+  }
+
+  public void setBucket(long bucket)
+  {
+    this.bucket = bucket;
+  }
+
+  public void setIdentifier(String identifier)
+  {
+    this.identifier = identifier;
+  }
+
+  public void setWindowSerde(Serde<Window, Slice> windowSerde)
+  {
+    this.windowSerde = windowSerde;
+  }
+
+  public void setWindowKeyPairSerde(Serde<ImmutablePair<Window, K>, Slice> windowKeyPairSerde)
+  {
+    this.windowKeyPairSerde = windowKeyPairSerde;
+  }
+
+  public void setValueSerde(Serde<V, Slice> valueSerde)
+  {
+    this.valueSerde = valueSerde;
+  }
+
+  @Override
+  public boolean containsWindow(Window window)
+  {
+    return internKeys.containsKey(window);
+  }
+
+  @Override
+  public long size()
+  {
+    return internKeys.size();
+  }
+
+  @Override
+  public void remove(Window window)
+  {
+    List<K> keys = internKeys.get(window);
+    for (K key : keys) {
+      internValues.remove(new ImmutablePair<>(window, key));
+    }
+    internKeys.removeAll(window);
+  }
+
+  @Override
+  public void migrateWindow(Window fromWindow, Window toWindow)
+  {
+    List<K> keys = internKeys.get(fromWindow);
+    internValues.remove(toWindow);
+    for (K key : keys) {
+      internKeys.put(toWindow, key);
+      ImmutablePair<Window, K> oldKey = new ImmutablePair<>(fromWindow, key);
+      ImmutablePair<Window, K> newKey = new ImmutablePair<>(toWindow, key);
+
+      V value = internValues.get(oldKey);
+      internValues.remove(oldKey);
+      internValues.put(newKey, value);
+    }
+    internKeys.removeAll(fromWindow);
+  }
+
+
+  @Override
+  public void setup(Context.OperatorContext context)
+  {
+    if (store == null) {
+      // provide a default managed state store
+      store = new ManagedStateSpillableStateStore();
+    }
+    sccImpl = new SpillableComplexComponentImpl(store);
+    internValues = sccImpl.newSpillableByteMap((identifier + "#values").getBytes(), bucket, windowKeyPairSerde, valueSerde);
```

Comment:
There is a constraint on the identifiers right now that they have to be the same number of bytes in length. Also you want the identifiers to be small since all the keys stored in managed state will be prefixed with it. For this case you can not specify an identifier. If you don't specify an identifier a single byte identifier will automatically generated for each data structure.